js/dactylo.js

```diff
@@ -99,20 +99,13 @@ window.addEventListener('DOMContentLoaded', () => {
   };
 
   const showKeys = () => {
-    gKeyList.innerHTML = '';
-    let level = 0;
-    for (let char of g30Keys.map(key => gKeyLayout.keymap[key][0])) {
-      if (char.length > 1) {
-        if (char = "**") {
-          char = "★";
-        } else {
-          char = char.slice(1);
-        }
-      }
-      const className = gKeyList.children.length < gLessonLevel ? '' : 'inactive';
-      gKeyList.innerHTML +=
-        `<kbd class='${className}' data-level="${++level}">${char}</kbd>`;
-    }
+    gKeyList.innerHTML = g30Keys
+      .map(key => gKeyLayout.keymap[key][0])
+      .map(char => char === '**' ? '★' : char) // 1dk/odk a.k.a. OneDeadKey
+      .map(char => char.length > 1 ? char.slice(1) : char) // other dead key
+      .map((char, idx) => `<kbd data-level="${idx + 1}"
+        class="${idx < gLessonLevel ? '' : 'inactive'}">${char}</kbd>`)
+      .join('');
```

Comment:
This is kind of dumb. If you’re chaining the same function (like `map`, here) multiple times in a row, you can (and probably should) move all of those steps into it’s own closure, and pass that as an argument to your `array.map(…)`. When reading your code later, you usually don’t really care what *exactly* all of the intermediate steps are (unless you’re debugging), but you do care about what the code means, and it’s pretty hard to see what is happening here.

It’s kind of like testing multiple boolean expressions in a row with `if`s :

```js
// This is stupid
if (condition1) {
  if (condition2) {
    // Code...
  }
}

// This is better
if (condition1 && condition2) {
  // Code...
}

// This is the best
const validInput = condition1 && condition2;
if (validInput) {
  // Code...
}
```

Here, if you just move all of the steps into it’s own closure, you get a similar effect : 

```js
const showKeys = () => {
  // Move all of the steps into it’s own closure, with an explicit name
  const serializeKey = (key, idx) => {
    let char = gKeyLayout.keymap[key][0];
    char = char === '**' ? '★' : char;
    char = char.length > 1 ? char.slice(1) : char;
    return `<kbd data-level="${idx + 1}"
      class="${idx < gLessonLevel ? '' : 'inactive'}">${char}</kbd>`;
  };

  // The code you actually care about becomes extremely simple to read
  gKeyList.innerHTML = g30Keys.map(serializeKey).join('');
};
```

And while you’re at it, since all of the code is moved to it’s own closure you have a lot more freedom when it comes to coding those intermediate steps : 

```js
  const showKeys = () => {
    const serializeKey = (key, idx) => {
        const char = gKeyLayout.keymap[key][0]
            .replace('**', '★')  // 1dk/odk a.k.a OneDeadKey
            .replace('*', '');   // other dead key, remove leading '*'

        const charClass = idx < gLessonLevel ? '' : 'class="inactive"';
        return `<kbd data-level="${idx + 1}" ${charClass}>${char}</kbd>`
    };

    gKeyList.innerHTML = g30Keys.map(serializeKey).join('');
  };
```

Since I’m already writing a fucking novel in this code review, I may as well give you some rule of thumbs I find helpful : 
1. If you have a single, simple expression in your `map`, `filter` (or whatever), you usually don’t need to move it to it’s own closure, it’s fine as is.
2. No nested closures. If you are moving code to it’s own closure, and start to need to move some of that code into another rested closure, you’re probably writing something way to complex and is a sign that your function has too many responsibilities, or that you’re missing a few valuable functions in your API.

Reply:
Fair enough. I won’t use exactly your code but I’ll avoid chaining `.map` calls.